src/coreclr/jit/morph.cpp

```diff
@@ -9273,6 +9296,127 @@ GenTree* Compiler::fgMorphCall(GenTreeCall* call)
     return call;
 }
 
+/*****************************************************************************
+ *
+ *  Expand and return the call target address for a VirtualCall
+ *  The code here should match that generated by LowerVirtualVtableCall
+ */
+
+GenTree* Compiler::fgExpandVirtualVtableCallTarget(GenTreeCall* call)
+{
+    GenTree* result;
+
+    JITDUMP("Expanding virtual call target for %d.%s:\n", call->gtTreeID, GenTree::OpName(call->gtOper));
+
+    noway_assert(call->gtCallType == CT_USER_FUNC);
+
+// get a reference to the thisPtr being passed
+#if 0
+    GenTree* thisPtr = call->gtCallThisArg->GetNode();
+#else
+    fgArgTabEntry* thisArgTabEntry = gtArgEntryByArgNum(call, 0);
+    GenTree*       thisPtr         = thisArgTabEntry->GetNode();
+#endif
+
+    // Make a copy of the thisPtr by cloning
+
+    assert(thisPtr->OperIsLocal()); // fgMorphArgs must enforce this invariant
+    thisPtr = gtClone(thisPtr, true);
+
+    noway_assert(thisPtr != nullptr);
+
+    // Get hold of the vtable offset (note: this might be expensive)
+    unsigned vtabOffsOfIndirection;
+    unsigned vtabOffsAfterIndirection;
+    bool     isRelative;
+    info.compCompHnd->getMethodVTableOffset(call->gtCallMethHnd, &vtabOffsOfIndirection, &vtabOffsAfterIndirection,
+                                            &isRelative);
+
+    // Dereference the this pointer to obtain the method table, it is called vtab below
+    GenTree* vtab;
+    assert(VPTR_OFFS == 0); // We have to add this value to the thisPtr to get the methodTable
+    vtab = gtNewOperNode(GT_IND, TYP_I_IMPL, thisPtr);
+    vtab->gtFlags |= GTF_IND_NONFAULTING;
+    vtab->gtFlags |= GTF_IND_INVARIANT;
+
+    // Get the appropriate vtable chunk
+    if (vtabOffsOfIndirection != CORINFO_VIRTUALCALL_NO_CHUNK)
+    {
+        if (isRelative)
+        {
+            // MethodTable offset is a relative pointer.
+            //
+            // Additional temporary variable is used to store virtual table pointer.
+            // Address of method is obtained by the next computations:
+            //
+            // Save relative offset to tmp (vtab is virtual table pointer, vtabOffsOfIndirection is offset of
+            // vtable-1st-level-indirection):
+            // tmp = vtab
+            //
+            // Save address of method to result (vtabOffsAfterIndirection is offset of vtable-2nd-level-indirection):
+            // result = [tmp + vtabOffsOfIndirection + vtabOffsAfterIndirection + [tmp + vtabOffsOfIndirection]]
+            //
+            //
+            // If relative pointers are also in second level indirection, additional temporary is used:
+            // var1 = vtab
+            // var2 = var1 + vtabOffsOfIndirection + vtabOffsAfterIndirection + [var1 + vtabOffsOfIndirection]
+            // result = [var2] + var2
+            //
+            unsigned varNum1 = lvaGrabTemp(true DEBUGARG("var1 - vtab"));
```

Comment:
Would be nice to have a bit more descriptive name for these temps, eg "first-level vtable", "second-level vtable"

Reply:
Is that what they mean? it wasn't clear to me that was their meanings.

Reply:
Maybe the first one could be called the "indirection slot address", and the second one the "slot address"?

See eg:

https://github.com/dotnet/runtime/blob/148bc0820805c7db4ff91d76f3a07bbecc43f160/src/coreclr/vm/methodtable.h#L1399-L1429

Reply:
I'd like to get rid of the unused isRelative code path, 
I'm really not OK with changing these names here when I don't really understand the implementation 
and I didn't see the implementation (or documentation of it) in the VM directory.

Reply:
@AndyAyersMS 
The comment in methodtable.h  that you linked to is what we use when isRelative is false.

It results in three indirections